codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/PresignerGenerator.kt

```diff
@@ -0,0 +1,91 @@
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.swift.codegen.ClientRuntimeTypes.Http.SdkHttpRequest
+import software.amazon.smithy.swift.codegen.ClientRuntimeTypes.Middleware.NoopHandler
+import software.amazon.smithy.swift.codegen.SwiftDelegator
+import software.amazon.smithy.swift.codegen.SwiftWriter
+import software.amazon.smithy.swift.codegen.core.CodegenContext
+import software.amazon.smithy.swift.codegen.core.toProtocolGenerationContext
+import software.amazon.smithy.swift.codegen.integration.MiddlewareExecutionGenerator
+import software.amazon.smithy.swift.codegen.integration.SwiftIntegration
+import software.amazon.smithy.swift.codegen.model.expectShape
+
+data class PresignableOperation(
+    val serviceId: String,
+    val operationId: String,
+    // TODO ~ Implementation of embedded presigned URLs is TBD
+    // val presignedParameterId: String?
+)
+
+class PresignerGenerator : SwiftIntegration {
+    override fun writeAdditionalFiles(ctx: CodegenContext, delegator: SwiftDelegator) {
+        val service = ctx.model.expectShape<ServiceShape>(ctx.settings.service)
+
+        if (!AWSSigningMiddleware.isSupportedAuthentication(ctx.model, service)) return
+        val presignOperations = service.allOperations
+            .map { ctx.model.expectShape<OperationShape>(it) }
+            .filter { operationShape -> operationShape.hasTrait(Presignable.ID) }
+            .map { operationShape ->
+                check(AWSSigningMiddleware.hasSigV4AuthScheme(ctx.model, service, operationShape)) { "Operation does not have valid auth trait" }
+                PresignableOperation(service.id.toString(), operationShape.id.toString())
+            }
+        presignOperations.forEach { presignableOperation ->
+            val op = ctx.model.expectShape<OperationShape>(presignableOperation.operationId)
+            val inputType = op.input.get().getName()
+            delegator.useFileWriter("${ctx.settings.moduleName}/models/$inputType+Presigner.swift") { writer ->
+                renderPresigner(writer, ctx, delegator, presignableOperation)
+            }
+        }
+    }
+
+    private fun renderPresigner(
+        writer: SwiftWriter,
+        ctx: CodegenContext,
+        delegator: SwiftDelegator,
+        presignableOperation: PresignableOperation
+    ) {
+        val serviceShape = ctx.model.expectShape<ServiceShape>(ctx.settings.service)
+        val protocolGenerator = ctx.protocolGenerator?.let { it } ?: run { return }
+        val protocolGeneratorContext = ctx.toProtocolGenerationContext(serviceShape, delegator)?.let { it } ?: run { return }
+
+        writer.addImport(AWSClientConfiguration)
+        writer.addImport(SdkHttpRequest)
+
+        val operationsIndex = OperationIndex.of(ctx.model)
+        val httpBindingResolver = protocolGenerator.getProtocolHttpBindingResolver(protocolGeneratorContext)
+
+        val op = ctx.model.expectShape<OperationShape>(presignableOperation.operationId)
```

Comment:
seems that you already have this value above, can you just pass it to the function on line 42 instead of calling it again?

Reply:
done